### PR TITLE
🍰 1054 blocked users

### LINKED ISSUE #1054: 🚀 [Feature] Block User
> [datenbrei](https://github.com/datenbrei) **Authored by [datenbrei](https://github.com/datenbrei)**
_<time datetime="2019-07-16T13:42:59Z" title="Tuesday, July 16th 2019, 3:42:59 pm +02:00">Jul 16, 2019</time>_
_Closed <time datetime="2019-08-14T14:36:41Z" title="Wednesday, August 14th 2019, 4:36:41 pm +02:00">Aug 14, 2019</time>_
---

## :rocket: [Feature] Block User

~~~gherkin
Feature: As a user
I'd like to block another user
to hinder him to see and interact with my contributions, profile and comments
and also that I don't see his profile, contributions and comments any longer.
~~~

### Design & Layout

block-user
blocked-users

### Additional context

This is part of Feature Blacklist #279 and a working out of #222. It should also be possible to unblock a user again later the same way.

### PR DESCRIPTION
**Pull request by Tirokk**
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-07T13:12:43Z" title="Wednesday, August 7th 2019, 3:12:43 pm +02:00">Aug 7, 2019</time>_
_Merged <time datetime="2019-08-14T14:36:41Z" title="Wednesday, August 14th 2019, 4:36:41 pm +02:00">Aug 14, 2019</time>_
---

### Issues

- fixes #1054 

### Todo

- [X] None

### CODE REVIEW

File: backend/src/schema/resolvers/users/blockedUsers.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/users/blockedUsers.spec.js:18]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-14T08:24:56Z" title="Wednesday, August 14th 2019, 10:24:56 am +02:00">Aug 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,393 @@
+import { createTestClient } from 'apollo-server-testing'
+import createServer from '../../../server'
+import Factory from '../../../seed/factories'
+import { gql } from '../../../jest/helpers'
+import { neode, getDriver } from '../../../bootstrap/neo4j'
+
+const driver = getDriver()
+const factory = Factory()
+const instance = neode()
+
+let currentUser
+let blockedUser
+let authenticatedUser
+let server
+
+beforeEach(() => {
+  authenticatedUser = undefined
+  ;({ server } = createServer({
```


I never saw this notation before.
What does the `;` mean? @roschaefer

File: webapp/locales/de.json
Comment (Tirokk) [inline webapp/locales/de.json:205]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-14T08:44:37Z" title="Wednesday, August 14th 2019, 10:44:37 am +02:00">Aug 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -189,6 +189,25 @@
       "submit": "Link hinzufügen",
       "successAdd": "Social-Media hinzugefügt. Profil aktualisiert!",
       "successDelete": "Social-Media gelöscht. Profil aktualisiert!"
+    },
+    "blocked-users": {
+      "name": "Blockierte Benutzer",
+      "explanation": {
+        "intro": "Wenn ein anderer Benutzer von dir blockiert wurde, dann passiert folgendes:",
+        "your-perspective": "In deiner Beitragsübersicht tauchen keine Beiträge der blockierten Person mehr auf.",
+        "their-perspective": "Umgekehrt das gleiche: Die blockierte Person sieht deine Beiträge auch nicht mehr in ihrer Übersicht.",
+        "search": "Die Beiträge von blockierten Personen verschwinden aus deinen Suchergebnissen.",
+        "notifications": "Von dir blockierte Personen erhalten keine Benachrichtigungen mehr, wenn sie in deinen Beiträgen erwähnt werden.",
+        "closing": "Das sollte fürs Erste genügen, damit blockierte Benutzer dich nicht mehr länger belästigen können."
+      },
+      "columns": {
+        "name": "Name",
+        "slug": "Alias"
```


Why here in German `Alias` and in English `Slug`? @roschaefer 

In the Admins Users list in both languages the slug is name `Slug`.
`Alias`is better understandable for the users for sure!

File: webapp/locales/en.json
Comment (Tirokk) [inline webapp/locales/en.json:205]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-14T08:46:49Z" title="Wednesday, August 14th 2019, 10:46:49 am +02:00">Aug 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -189,6 +189,26 @@
       "submit": "Add link",
       "successAdd": "Added social media. Updated user profile!",
       "successDelete": "Deleted social media. Updated user profile!"
+    },
+    "blocked-users": {
+      "name": "Blocked users",
+      "explanation": {
+        "intro": "If another user has been blocked by you, this is what happens:",
+        "your-perspective": "The blocked person's posts will no longer appear in your news feed.",
+        "their-perspective": "Vice versa: The blocked person will also no longer see your posts in their news feed.",
+        "search": "Posts of blocked people disappear from your search results.",
+        "notifications": "Blocked users will no longer receive notifications if they are mentioned in your posts.",
+        "closing": "This should be sufficient for now so that blocked users can no longer bother you."
+
+      },
+      "columns": {
+        "name": "Name",
+        "slug": "Slug"
```


See above …

File: webapp/locales/de.json
Comment (Tirokk) [reply webapp/locales/de.json:205]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-14T08:57:40Z" title="Wednesday, August 14th 2019, 10:57:40 am +02:00">Aug 14, 2019</time>_

---


I thought that "Slug" is not so common in German for most people to understand. I don't know what it actually translates to, I believe it's "Schnecke" and refers to the pattern of strings that have no witespaces as delimiter (in a URL).

Good luck understanding that one..

File: backend/src/schema/resolvers/users/blockedUsers.spec.js
Comment (Tirokk) [reply backend/src/schema/resolvers/users/blockedUsers.spec.js:18]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-14T08:59:49Z" title="Wednesday, August 14th 2019, 10:59:49 am +02:00">Aug 14, 2019</time>_

---


It was added by eslint. The problem is that if you have e.g.
```js
const { a } = something
```
is valid and no problem but
```js
let a
{ a } = something
```
then babel gets confused. You need to wrap it with parantheses and apparently it's good practice (?) to add a semicolon as delimiter?

**Comment by Tirokk** [comment]
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-14T09:02:49Z" title="Wednesday, August 14th 2019, 11:02:49 am +02:00">Aug 14, 2019</time>_

---


@Tirokk forgot to add 31299235a3850e47ed885cf71579c3117da52dac could you recheck, please?

File: backend/src/schema/resolvers/users/blockedUsers.spec.js
Comment (Tirokk) [reply backend/src/schema/resolvers/users/blockedUsers.spec.js:18]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-14T09:04:48Z" title="Wednesday, August 14th 2019, 11:04:48 am +02:00">Aug 14, 2019</time>_

---


Ah, okay.

File: webapp/locales/de.json
Comment (Tirokk) [reply webapp/locales/de.json:205]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-14T09:13:42Z" title="Wednesday, August 14th 2019, 11:13:42 am +02:00">Aug 14, 2019</time>_

---


No, I think "Slug" is also a bad name. Shall we use "alias" everywhere?

**Comment by Tirokk** [comment]
> [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-14T09:22:00Z" title="Wednesday, August 14th 2019, 11:22:00 am +02:00">Aug 14, 2019</time>_

---


DE: Nun funzt es. Wir fliegen zu den Sternen …
EN: Now it’s functioning. We flying to the stars …

Cool thing @roschaefer !!! ❄️💫🚀🚀🚀

Wisdom

**Comment by Tirokk** [comment]
> [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-14T09:22:46Z" title="Wednesday, August 14th 2019, 11:22:46 am +02:00">Aug 14, 2019</time>_

---


Love to see what @mattwr18 says about the `Slug` …

File: webapp/locales/de.json
Comment (Tirokk) [reply webapp/locales/de.json:205]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-14T10:34:33Z" title="Wednesday, August 14th 2019, 12:34:33 pm +02:00">Aug 14, 2019</time>_

---


I understand what a slug is, but can't remember when I gained that understanding :woman_shrugging: 
An alias, I guess is better understood by the general user, but it's not really the same, is it?

File: backend/src/schema/resolvers/posts.js
Comment (Tirokk) [inline backend/src/schema/resolvers/posts.js:13]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-14T10:48:27Z" title="Wednesday, August 14th 2019, 12:48:27 pm +02:00">Aug 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,7 +1,74 @@
 import uuid from 'uuid/v4'
+import { neo4jgraphql } from 'neo4j-graphql-js'
 import fileUpload from './fileUpload'
+import { getBlockedUsers, getBlockedByUsers } from './users.js'
+import { mergeWith, isArray } from 'lodash'
+
+const filterForBlockedUsers = async (params, context) => {
+  if (!context.user) return params
+  const [blockedUsers, blockedByUsers] = await Promise.all([
+    getBlockedUsers(context),
+    getBlockedByUsers(context),
+  ])
+  const badIds = [...blockedByUsers.map(b => b.id), ...blockedUsers.map(b => b.id)]
```


what do you mean `badIds`? these are the ids' of users that have been blocked by a currentUser, oder?

File: backend/src/schema/resolvers/users.js
Comment (Tirokk) [inline backend/src/schema/resolvers/users.js:20]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-14T10:49:24Z" title="Wednesday, August 14th 2019, 12:49:24 pm +02:00">Aug 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -6,8 +6,45 @@ import Resolver from './helpers/Resolver'
 
 const instance = neode()
 
+export const getBlockedUsers = async context => {
+  const { neode } = context
+  const userModel = neode.model('User')
+  let blockedUsers = neode
+    .query()
+    .match('user', userModel)
+    .where('user.id', context.user.id)
+    .relationship(userModel.relationships().get('blocked'))
+    .to('blocked', userModel)
+    .return('blocked')
+  blockedUsers = await blockedUsers.execute()
+  blockedUsers = blockedUsers.records.map(r => r.get('blocked').properties)
```


neode magic!!

File: backend/src/schema/resolvers/users/blockedUsers.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/users/blockedUsers.spec.js:50]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-14T10:50:38Z" title="Wednesday, August 14th 2019, 12:50:38 pm +02:00">Aug 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,393 @@
+import { createTestClient } from 'apollo-server-testing'
+import createServer from '../../../server'
+import Factory from '../../../seed/factories'
+import { gql } from '../../../jest/helpers'
+import { neode, getDriver } from '../../../bootstrap/neo4j'
+
+const driver = getDriver()
+const factory = Factory()
+const instance = neode()
+
+let currentUser
+let blockedUser
+let authenticatedUser
+let server
+
+beforeEach(() => {
+  authenticatedUser = undefined
+  ;({ server } = createServer({
+    context: () => {
+      return {
+        user: authenticatedUser,
+        driver,
+        neode: instance,
+        cypherParams: {
+          currentUserId: authenticatedUser ? authenticatedUser.id : null,
+        },
+      }
+    },
+  }))
+})
+
+afterEach(async () => {
+  await factory.cleanDatabase()
+})
+
+describe('blockedUsers', () => {
+  let blockedUserQuery
+  beforeEach(() => {
+    blockedUserQuery = gql`
+      query {
+        blockedUsers {
+          id
+          name
+          isBlocked
+        }
+      }
+    `
+  })
+
+  it('throws permission error', async () => {
```


this is describing `unauthenticated`, correct?

File: backend/src/schema/resolvers/users/blockedUsers.spec.js
Comment (Tirokk) [inline backend/src/schema/resolvers/users/blockedUsers.spec.js:94]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-14T10:55:57Z" title="Wednesday, August 14th 2019, 12:55:57 pm +02:00">Aug 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,393 @@
+import { createTestClient } from 'apollo-server-testing'
+import createServer from '../../../server'
+import Factory from '../../../seed/factories'
+import { gql } from '../../../jest/helpers'
+import { neode, getDriver } from '../../../bootstrap/neo4j'
+
+const driver = getDriver()
+const factory = Factory()
+const instance = neode()
+
+let currentUser
+let blockedUser
+let authenticatedUser
+let server
+
+beforeEach(() => {
+  authenticatedUser = undefined
+  ;({ server } = createServer({
+    context: () => {
+      return {
+        user: authenticatedUser,
+        driver,
+        neode: instance,
+        cypherParams: {
+          currentUserId: authenticatedUser ? authenticatedUser.id : null,
+        },
+      }
+    },
+  }))
+})
+
+afterEach(async () => {
+  await factory.cleanDatabase()
+})
+
+describe('blockedUsers', () => {
+  let blockedUserQuery
+  beforeEach(() => {
+    blockedUserQuery = gql`
+      query {
+        blockedUsers {
+          id
+          name
+          isBlocked
+        }
+      }
+    `
+  })
+
+  it('throws permission error', async () => {
+    const { query } = createTestClient(server)
+    const result = await query({ query: blockedUserQuery })
+    expect(result.errors[0]).toHaveProperty('message', 'Not Authorised!')
+  })
+
+  describe('authenticated and given a blocked user', () => {
+    beforeEach(async () => {
+      currentUser = await instance.create('User', {
+        name: 'Current User',
+        id: 'u1',
+      })
+      blockedUser = await instance.create('User', {
+        name: 'Blocked User',
+        id: 'u2',
+      })
+      await currentUser.relateTo(blockedUser, 'blocked')
+      authenticatedUser = await currentUser.toJson()
+    })
+
+    it('returns a list of blocked users', async () => {
+      const { query } = createTestClient(server)
+      await expect(query({ query: blockedUserQuery })).resolves.toEqual(
+        expect.objectContaining({
+          data: {
+            blockedUsers: [
+              {
+                name: 'Blocked User',
+                id: 'u2',
+                isBlocked: true,
+              },
+            ],
+          },
+        }),
+      )
+    })
+  })
+})
+
+describe('block', () => {
+  let blockAction
+
+  beforeEach(() => {
+    currentUser = undefined
+    blockAction = variables => {
```


I appreciate you making it clear what the variables are doing! :blush:

File: webapp/locales/en.json
Comment (Tirokk) [inline webapp/locales/en.json:211]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-14T11:06:12Z" title="Wednesday, August 14th 2019, 1:06:12 pm +02:00">Aug 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -189,6 +189,26 @@
       "submit": "Add link",
       "successAdd": "Added social media. Updated user profile!",
       "successDelete": "Deleted social media. Updated user profile!"
+    },
+    "blocked-users": {
+      "name": "Blocked users",
+      "explanation": {
+        "intro": "If another user has been blocked by you, this is what happens:",
+        "your-perspective": "The blocked person's posts will no longer appear in your news feed.",
+        "their-perspective": "Vice versa: The blocked person will also no longer see your posts in their news feed.",
+        "search": "Posts of blocked people disappear from your search results.",
+        "notifications": "Blocked users will no longer receive notifications if they are mentioned in your posts.",
+        "closing": "This should be sufficient for now so that blocked users can no longer bother you."
+
+      },
+      "columns": {
+        "name": "Name",
+        "slug": "Slug"
+      },
+      "empty": "So far, you have not blocked anybody.",
+      "how-to": "You can block other users on their profile page via the content menu.",
+      "block": "Block User",
+      "unblock": "Unblock User"
```


we have talked about capitalizing the first word and using lowercase for the following words.

File: webapp/pages/profile/_id/_slug.vue
Comment (Tirokk) [inline webapp/pages/profile/_id/_slug.vue:391]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-14T11:07:05Z" title="Wednesday, August 14th 2019, 1:07:05 pm +02:00">Aug 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -372,6 +380,16 @@ export default {
       }
       return this.uniq(this.Post.filter(post => !post.deleted))
     },
+    async block(user) {
+      await this.$apollo.mutate({ mutation: Block(), variables: { id: user.id } })
+      this.$apollo.queries.User.refetch()
+      this.$apollo.queries.Post.refetch()
+    },
+    async unblock(user) {
+      await this.$apollo.mutate({ mutation: Unblock(), variables: { id: user.id } })
+      this.$apollo.queries.User.refetch()
+      this.$apollo.queries.Post.refetch()
```


is there a reason we need to refetch the User and Post here? can't we update the cache on a successful mutation, or did you try that and it didn't work for some reason?

**Review by Mogge — CHANGES_REQUESTED** [review]
A masterpiece again, dear!!! 🚀💫
Wonderful! The users will thank you! @roschaefer 

When I block a user the blocked users list is not updated, if I go there immediately:
Bildschirmfoto 2019-08-14 um 10 17 44
Bildschirmfoto 2019-08-14 um 10 18 22

If I reload the page the list is correct.

The other way around, if I unblock a user, is the same:
Bildschirmfoto 2019-08-14 um 10 21 06
Bildschirmfoto 2019-08-14 um 10 09 15

The user is still listed and after reload everything is correct.
A caching problem?

**Review by Mogge — APPROVED** [review]

**Review by Mogge — APPROVED** [review]

**Review by Mogge — APPROVED** [review]
This is really great @roschaefer!!
I added some comments, but am approving this, let me know if they made sense.

It took me a little bit to review this cause there are quite a few changes in the tests that are related to refactoring to use `neode`, but I can understand that while adding the tests you wanted to use `neode` and it was a good time to refactor.

**Pull request closed without merging**
